hw/misc/hermes.c

```diff
@@ -977,7 +1006,13 @@ static void hermes_bar2_init_dir_reg(HermesState *hermes,
 {
     qemu_mutex_init(&dir->dma_mutex);
     qemu_cond_init(&dir->dma_cond);
-    qemu_thread_create(&dir->dma_thread, dir->name, dir->entry, hermes,
+
+    struct hermes_dir_pair pair = {
```

Comment:
We dont' need a new struct (plus this won't work because you're allocating on the stack and by the time the thread tries to use it might be gone).

How about something like this?
```diff
diff --git a/hw/misc/hermes.c b/hw/misc/hermes.c
index 41c1e6b643..3f49c5a7ca 100644
--- a/hw/misc/hermes.c
+++ b/hw/misc/hermes.c
@@ -299,15 +299,21 @@ static void hermes_dma_thread(HermesState *hermes,
 
 static void *hermes_h2c_dma_thread(void *opaque)
 {
-    struct hermes_dir_pair *pair = opaque;
-    hermes_dma_thread(pair->hermes, &pair->hermes->bar2.h2c[pair->dir->ch]);
+    struct hermes_bar2_dir_reg *dir = opaque;
+    struct hermes_bar2 *bar2 = container_of(dir, struct hermes_bar2, h2c[dir->ch]);
+    HermesState *hermes = container_of(bar2, HermesState, bar2);
+
+    hermes_dma_thread(hermes, dir);
     return NULL;
 }
 
 static void *hermes_c2h_dma_thread(void *opaque)
 {
-    struct hermes_dir_pair *pair = opaque;
-    hermes_dma_thread(pair->hermes, &pair->hermes->bar2.c2h[pair->dir->ch]);
+    struct hermes_bar2_dir_reg *dir = opaque;
+    struct hermes_bar2 *bar2 = container_of(dir, struct hermes_bar2, c2h[dir->ch]);
+    HermesState *hermes = container_of(bar2, HermesState, bar2);
+
+    hermes_dma_thread(hermes, dir);
     return NULL;
 }
 
@@ -1007,12 +1013,7 @@ static void hermes_bar2_init_dir_reg(HermesState *hermes,
     qemu_mutex_init(&dir->dma_mutex);
     qemu_cond_init(&dir->dma_cond);
 
-    struct hermes_dir_pair pair = {
-        .hermes = hermes,
-        .dir = dir,
-    };
-
-    qemu_thread_create(&dir->dma_thread, dir->name, dir->entry, &pair,
+    qemu_thread_create(&dir->dma_thread, dir->name, dir->entry, dir,
                        QEMU_THREAD_JOINABLE);
 }
```

Reply:
Good catch. Thanks.